src/dot.cc

```diff
@@ -5,246 +5,293 @@
-    blas_error_if( n < 0 );      // standard BLAS returns, doesn't fail
-    blas_error_if( incx == 0 );  // standard BLAS doesn't detect inc[xy] == 0
-    blas_error_if( incy == 0 );
-
-    // check for overflow in native BLAS integer type, if smaller than int64_t
-    if (sizeof(int64_t) > sizeof(blas_int)) {
-        blas_error_if( n              > std::numeric_limits<blas_int>::max() );
-        blas_error_if( std::abs(incx) > std::numeric_limits<blas_int>::max() );
-        blas_error_if( std::abs(incy) > std::numeric_limits<blas_int>::max() );
-    }
-
-    blas_int n_    = (blas_int) n;
-    blas_int incx_ = (blas_int) incx;
-    blas_int incy_ = (blas_int) incy;
-    return BLAS_ddot( &n_, x, &incx_, y, &incy_ );
+    return BLAS_ddot( &n, x, &incx, y, &incy );
 }
 
-// -----------------------------------------------------------------------------
-/// @ingroup dot
-std::complex<float> dot(
-    int64_t n,
-    std::complex<float> const *x, int64_t incx,
-    std::complex<float> const *y, int64_t incy )
+//------------------------------------------------------------------------------
+/// Low-level overload wrapper calls Fortran, complex<float> version.
+/// @ingroup dot_internal
+inline std::complex<float> dot(
+    blas_int n,
+    std::complex<float> const* x, blas_int incx,
+    std::complex<float> const* y, blas_int incy )
 {
-    // check arguments
-    blas_error_if( n < 0 );      // standard BLAS returns, doesn't fail
-    blas_error_if( incx == 0 );  // standard BLAS doesn't detect inc[xy] == 0
-    blas_error_if( incy == 0 );
-
-    // check for overflow in native BLAS integer type, if smaller than int64_t
-    if (sizeof(int64_t) > sizeof(blas_int)) {
-        blas_error_if( n              > std::numeric_limits<blas_int>::max() );
-        blas_error_if( std::abs(incx) > std::numeric_limits<blas_int>::max() );
-        blas_error_if( std::abs(incy) > std::numeric_limits<blas_int>::max() );
-    }
-
-    blas_int n_    = (blas_int) n;
-    blas_int incx_ = (blas_int) incx;
-    blas_int incy_ = (blas_int) incy;
-
     #ifdef BLAS_COMPLEX_RETURN_ARGUMENT
         // Intel icc convention
         std::complex<float> value;
-        BLAS_cdotc( (blas_complex_float*) &value, &n_,
-                    (blas_complex_float*) x, &incx_,
-                    (blas_complex_float*) y, &incy_ );
+        BLAS_cdotc( (blas_complex_float*) &value, &n,
+                    (blas_complex_float*) x, &incx,
+                    (blas_complex_float*) y, &incy );
         return value;
     #else
         // GNU gcc convention
         blas_complex_float value
-             = BLAS_cdotc( &n_,
-                           (blas_complex_float*) x, &incx_,
-                           (blas_complex_float*) y, &incy_ );
+             = BLAS_cdotc( &n,
+                           (blas_complex_float*) x, &incx,
+                           (blas_complex_float*) y, &incy );
         return *reinterpret_cast< std::complex<float>* >( &value );
     #endif
```

Comment:
This structure is duplicated multiple times. Could we make a wrapper? Like:
```
template <scalar_t, blas_t, f>
scalar_t dot( n, x, incx, y, incy ) {
//Contains the ifdef and reinterpret_Cast, and call the function f
[...]
}
```

Reply:
I see. It might require too much effort for now. But I thought, by passing `f` as a template it would work.